lib/utils/aws/organizations/matchingaccounts.go

```diff
@@ -0,0 +1,300 @@
+	organizations.ListAccountsAPIClient
+	organizations.ListAccountsForParentAPIClient
+}
+
+type awsOrgItem struct {
+	id                  string
+	organizationalUnits []*awsOrgItem
+	accounts            []string
+}
+
+// MatchingAccountsFilter defines the filter to apply when retrieving matching accounts from an AWS Organization.
+type MatchingAccountsFilter struct {
+	// OrganizationID is the ID of the AWS Organization to query.
+	// Required.
+	OrganizationID string
+
+	// IncludeOUs is the list of Organizational Unit IDs to include.
+	// If empty or contains "*", all OUs are included (except those in ExcludeOUs).
+	// If contains specific OU IDs, only those OUs and their children OUs are included.
+	// Optional.
+	IncludeOUs []string
+
+	// ExcludeOUs is the list of Organizational Unit IDs to exclude.
+	// If contains "*", all OUs are excluded.
+	// Optional.
+	ExcludeOUs []string
+}
+
+func (m *MatchingAccountsFilter) checkAndSetDefaults() error {
+	if m.OrganizationID == "" {
+		return trace.BadParameter("OrganizationID is required")
+	}
+
+	if len(m.IncludeOUs) == 0 {
+		return trace.BadParameter("at least one Organizational Unit must be included")
+	}
+
+	if len(m.IncludeOUs) > 1 && slices.Contains(m.IncludeOUs, allOrganizationalUnits) {
+		return trace.BadParameter("IncludeOUs cannot contain '*' along with other OU IDs")
+	}
+
+	if len(m.ExcludeOUs) > 1 && slices.Contains(m.ExcludeOUs, allOrganizationalUnits) {
+		return trace.BadParameter("ExcludeOUs cannot contain '*' along with other OU IDs")
+	}
+
+	return nil
+}
+
+func (m *MatchingAccountsFilter) isIncludeNothing() bool {
+	return len(m.ExcludeOUs) == 1 && m.ExcludeOUs[0] == allOrganizationalUnits
+}
+
+func (m *MatchingAccountsFilter) isIncludeAll() bool {
+	return len(m.ExcludeOUs) == 0 && (len(m.IncludeOUs) == 0 || (len(m.IncludeOUs) == 1 && m.IncludeOUs[0] == allOrganizationalUnits))
+}
+
+// MatchingAccounts returns the list of account IDs that are part of the organization and match the filter.
+// Every OU in ExcludeOUs is excluded from the results, including its children OUs.
+func MatchingAccounts(ctx context.Context, orgsClient OrganizationsClient, filter MatchingAccountsFilter) ([]string, error) {
+	if err := filter.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if filter.isIncludeNothing() {
+		return []string{}, nil
+	}
+
+	if filter.isIncludeAll() {
+		accountIDs, err := allAccounts(ctx, orgsClient, filter.OrganizationID)
+		return accountIDs, trace.Wrap(err)
+	}
+
+	// General case: build the organization tree, this tree already removes all the excluded OUs and their children OUs.
+	orgTree, err := buildOrgTree(ctx, orgsClient, filter)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Then: iterate over the tree and collect the included accounts.
+	includeEverything := len(filter.IncludeOUs) == 0 || filter.IncludeOUs[0] == allOrganizationalUnits
+	return collectIncludedAccounts(orgTree, filter.IncludeOUs, includeEverything), nil
```

Comment:
```suggestion
	includeParent := len(filter.IncludeOUs) == 1 && filter.IncludeOUs[0] == allOrganizationalUnits
	return collectIncludedAccounts(orgTree, filter.IncludeOUs, includeParent), nil
```

although I feel like this is still wrong since we don't check excludes here.
Should we even check this? It seems like the include all case is handled above with an early return if `filter.isIncludeAll()` is true

Reply:
I'll rename this to `includeRootAndChildrenAccounts`
This variables is meant to say that the account ids of the root OU and its children must be included.

It's not quite the same as the include all case.

----
If we are including everything, that means no exclude OU is defined.
In this case, an excluded OU as defined.

In this case, we build the org tree and exclude all the OUs and their children that are in the exclude list.